File: Session13/78. Subsets.java
class Solution {
    List<List<Integer>> ans=new ArrayList<List<Integer>>();
    public List<List<Integer>> subsets(int[] nums) {
        int n=nums.length;
        List<Integer> temp=new ArrayList<Integer>();
        backtrack(nums,temp,0,n);
        return ans;
    }
    void backtrack(int nums[],List<Integer> temp,int pos,int n)
    {
        if(pos==n)
        {
            ans.add(new ArrayList<Integer>(temp));
            return;
        }
        temp.add(nums[pos]);
        backtrack(nums,temp,pos+1,n);
        temp.remove(temp.size()-1);
        backtrack(nums,temp,pos+1,n);
    }
}

File: Session13/78. Subsets.py
class Solution:
    
#     def subsets(self, nums: List[int]) -> List[List[int]]:
#         ans=[]
#         n=len(nums)
#         self.helper(0,n,nums,[],ans)
#         return ans
    
#     def helper(self,pos,n,nums,temp,ans):
#         if pos>=n:
#             ans.append(temp[::])
#             return
#         temp.append(nums[pos])
#         self.helper(pos+1,n,nums,temp,ans)
#         temp.pop(-1)
#         self.helper(pos+1,n,nums,temp,ans)


#     def __init__(self):
#         self.ans=[]
    
#     def subsets(self, nums: List[int]) -> List[List[int]]:
#         n=len(nums)
#         self.helper(0,n,nums,[])
#         return self.ans
    
#     def helper(self,pos,n,nums,temp):
#         if pos>=n:
#             self.ans.append(temp[::])
#             return
#         temp.append(nums[pos])
#         self.helper(pos+1,n,nums,temp)
#         temp.pop(-1)
#         self.helper(pos+1,n,nums,temp)
    
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n=len(nums)
        ans = []
        def helper(pos,n,nums,temp):
            if pos>=n:
                ans.append(temp[::])
                return
            temp.append(nums[pos])
            helper(pos+1,n,nums,temp)
            temp.pop(-1)
            helper(pos+1,n,nums,temp)
        helper(0,n,nums,[])
        return ans
    